bigtable/client/table_bulk_apply_test.cc

```diff
@@ -0,0 +1,377 @@
+        {
+          auto &e = *r->add_entries();
+          e.set_index(0);
+          e.mutable_status()->set_code(grpc::OK);
+        }
+        return true;
+      }))
+      .WillOnce(Return(false));
+  EXPECT_CALL(*r1, Finish())
+      .WillOnce(Return(grpc::Status(grpc::StatusCode::ABORTED, "")));
+
+  auto create_cancelled_stream = [&](grpc::ClientContext *,
+                                     btproto::MutateRowsRequest const &) {
+    auto stream = absl::make_unique<MockReader>();
+    EXPECT_CALL(*stream, Read(_)).WillOnce(Return(false));
+    EXPECT_CALL(*stream, Finish())
+        .WillOnce(Return(grpc::Status(grpc::StatusCode::ABORTED, "")));
+    return stream.release();
+  };
+
+  EXPECT_CALL(*client.mock_stub, MutateRowsRaw(_, _))
+      .WillOnce(Invoke(
+          [&r1](grpc::ClientContext *, btproto::MutateRowsRequest const &) {
+            return r1.release();
+          }))
+      .WillOnce(Invoke(create_cancelled_stream))
+      .WillOnce(Invoke(create_cancelled_stream));
+
+  EXPECT_THROW(table.BulkApply(bt::BulkMutation(
+                   bt::SingleRowMutation(
+                       "foo", {bigtable::SetCell("fam", "col", 0, "baz")}),
+                   bt::SingleRowMutation(
+                       "bar", {bigtable::SetCell("fam", "col", 0, "qux")}))),
+               std::exception);
+}
+
+/// @test Verify that Table::BulkApply() retries only idempotent mutations.
+TEST(TableBulkApply, RetryOnlyIdempotent) {
+  MockClient client;
+  using namespace ::testing;
+  EXPECT_CALL(client, Stub())
+      .WillRepeatedly(Invoke(
+          [&client]() -> google::bigtable::v2::Bigtable::StubInterface & {
+            return *client.mock_stub;
+          }));
+  namespace btproto = ::google::bigtable::v2;
+
+  // We will send both idempotent and non-idempotent mutations.  We prepare the
+  // mocks to return an empty stream in the first RPC request.  That will force
+  // the client to only retry the idempotent mutations.
+  auto r1 = absl::make_unique<MockReader>();
+  EXPECT_CALL(*r1, Read(_)).WillOnce(Return(false));
+  EXPECT_CALL(*r1, Finish()).WillOnce(Return(grpc::Status::OK));
+
+  auto r2 = absl::make_unique<MockReader>();
+  EXPECT_CALL(*r2, Read(_))
+      .WillOnce(Invoke([](btproto::MutateRowsResponse *r) {
+        {
+          auto &e = *r->add_entries();
+          e.set_index(0);
+          e.mutable_status()->set_code(grpc::OK);
+        }
+        return true;
+      }))
+      .WillOnce(Return(false));
+  EXPECT_CALL(*r2, Finish()).WillOnce(Return(grpc::Status::OK));
+
+  EXPECT_CALL(*client.mock_stub, MutateRowsRaw(_, _))
+      .WillOnce(Invoke(
+          [&r1](grpc::ClientContext *, btproto::MutateRowsRequest const &) {
+            return r1.release();
+          }))
+      .WillOnce(Invoke(
+          [&r2](grpc::ClientContext *, btproto::MutateRowsRequest const &) {
+            return r2.release();
+          }));
+
+  namespace bt = bigtable;
+  bt::Table table(&client, "foo_table");
```

Comment:
`s/bt::Table/bigtable::Table/` ?

You're using the `bigtable` namespace below, so using `bt` as an abbreviation doesn't help much here.

Reply:
Fixed.